src/builtins/complete.cpp

```diff
@@ -422,6 +423,10 @@ maybe_t<int> builtin_complete(parser_t &parser, io_streams_t &streams, const wch
             flags |= COMPLETE_DONT_SORT;
         }
 
+        if (!escape) {
+            flags |= COMPLETE_DONT_ESCAPE;
```

Comment:
Problem is, if there is a space in the filename we will no longer escape it, so we'll insert incorrect completions again.
This is a difficult issue.

- We could add a hack that replaces a leading  `\~` with `~`. I don't like that because the backslash could be legit.
- Or just not use `__fish_complete_suffix`, but that would pessimate the common case.
- Or expand the ~ to the value of $HOME, as suggested in https://github.com/fish-shell/fish-shell/issues/4570#issuecomment-718082497

Neither of those are nice.
A real fix needs to add metadata to each completion entry, because the knowledge that `~` is actually `$HOME` is lost in serialization.
Sadly this is difficult to do without breakage.

Right now each completion is serialized as

    <completion text> \t <description>

we could extend that to

    <completion text> \t <description> \t <completion text but it's already escaped>

then, if fish sees a completion with the three-column format, we just the last column, and don't escape it.

A related issue is #3469: `complete -C'\\$HOME-$HOM'` outputs `$HOME-$HOME`.

Reply:
Ah, that's a good point. If I change this to `COMPLETE_DONT_ESCAPE_TILDES` instead, it fixes that bug. Doing so essentially treats the completions as regular files completions and escapes them as such. Which in this case is what we want because the completions are output by `__fish_complete_directories` / `__fish_complete_suffix` and correspond to files.

Reply:
Yep that would work until we come up with a more general solution.

I'd like to avoid updating all call sites of `__fish_complete_suffix` plus possibly others -- fish completions should remain easy to write -- so why don't we use `COMPLETE_DONT_ESCAPE_TILDES` always (for all things passed to `complete -a`)? Since `~` is only special at the beginning of a token, I expect few wrong completions. Is there any program that expects arguments that start with a literal `~`? Probably not because shells punish that :-)

Reply:
I like that idea. I just wasn't sure about the possible consequences. I'll go ahead and push a commit with that approach.

Reply:
merged as 4a3e55f69c63b6b46ae81e0ddc2e52326c072872 BTW the `if` was not necessary because the flags are only ever used for `-a` arguments anyway